MoveGroupInterface setJointValueTarget with Pose error: Returning dirty link transforms

### Description

Since I upgraded to moveit 1.1.12, setting the joint value target to a pose with a frame different from the original end effector (e.g. an attached camera) doesn't work anymore. I receive warnings about "Returning dirty link transforms", and then setting the target fails. This worked in the previous version (1.1.11), and if I downgrade the packages, it works again.
I originally use a setup with xarm_ros, but managed to reproduce it with the panda_moveit_config package, which should be easier to test.

Looking at the changelog, this is likely related to #3388.

### Your environment
* ROS Distro: Noetic
* OS Version:Ubuntu 20.04
* Binary build, Release version 1.1.12

### Steps to reproduce
Clone and build this package: https://github.com/Eruvae/mgr_bug_minimal
Launch panda_moveit_config demo: roslaunch panda_moveit_config demo.launch
Run the demo node: rosrun mgr_bug_minimal mgr_bug_minimal

### Expected behaviour
The node should successfully set the joint value target from the given pose. Output:
```
[ INFO] [1687969414.891391301]: Loading robot model 'panda'...
[ INFO] [1687969416.029813122]: Ready to take commands for planning group panda_arm.
[ INFO] [1687969416.282065594]: setJointValueTarget() succeeded
```
This works with MoveIT 1.1.11 installed.

### Actual behaviour
Setting the joint value target fails with warnings about dirty link transforms. Output:
```
[ INFO] [1687969509.832041115]: Loading robot model 'panda'...
[ INFO] [1687969510.902151232]: Ready to take commands for planning group panda_arm.
[ WARN] [1687969511.181761120]: Returning dirty link transforms
[ WARN] [1687969511.181847960]: Returning dirty link transforms
[ WARN] [1687969511.181884760]: Returning dirty link transforms
[ WARN] [1687969511.181911947]: Returning dirty link transforms
[ERROR] [1687969511.232013851]: setJointValueTarget() failed
```
### Installed MoveIT packages
These are my installed packages related to moveit with the issue:
```
ros-noetic-chomp-motion-planner/focal,focal,now 1.1.12-1focal.20230524.155215 amd64 [installed]
ros-noetic-moveit-commander/focal,focal,now 1.1.12-1focal.20230524.164353 amd64 [installed]
ros-noetic-moveit-core/focal,focal,now 1.1.12-1focal.20230524.152908 amd64 [installed]
ros-noetic-moveit-fake-controller-manager/focal,focal,now 1.1.12-1focal.20230524.161352 amd64 [installed]
ros-noetic-moveit-kinematics/focal,focal,now 1.1.12-1focal.20230524.161147 amd64 [installed]
ros-noetic-moveit-msgs/focal,focal,now 0.11.4-1focal.20230216.003351 amd64 [installed]
ros-noetic-moveit-planners-chomp/focal,focal,now 1.1.12-1focal.20230524.164352 amd64 [installed]
ros-noetic-moveit-planners-ompl/focal,focal,now 1.1.12-1focal.20230524.161205 amd64 [installed]
ros-noetic-moveit-planners/focal,focal,now 1.1.12-1focal.20230524.170420 amd64 [installed]
ros-noetic-moveit-plugins/focal,focal,now 1.1.12-1focal.20230524.162050 amd64 [installed]
ros-noetic-moveit-ros-benchmarks/focal,focal,now 1.1.12-1focal.20230524.162901 amd64 [installed]
ros-noetic-moveit-ros-control-interface/focal,focal,now 1.1.12-1focal.20230524.155610 amd64 [installed]
ros-noetic-moveit-ros-manipulation/focal,focal,now 1.1.12-1focal.20230524.163019 amd64 [installed,automatic]
ros-noetic-moveit-ros-move-group/focal,focal,now 1.1.12-1focal.20230524.161815 amd64 [installed]
ros-noetic-moveit-ros-occupancy-map-monitor/focal,focal,now 1.1.12-1focal.20230524.155332 amd64 [installed]
ros-noetic-moveit-ros-perception/focal,focal,now 1.1.12-1focal.20230524.161214 amd64 [installed]
ros-noetic-moveit-ros-planning-interface/focal,focal,now 1.1.12-1focal.20230524.163543 amd64 [installed]
ros-noetic-moveit-ros-planning/focal,focal,now 1.1.12-1focal.20230524.155710 amd64 [installed]
ros-noetic-moveit-ros-robot-interaction/focal,focal,now 1.1.12-1focal.20230524.161232 amd64 [installed]
ros-noetic-moveit-ros-visualization/focal,focal,now 1.1.12-1focal.20230524.164446 amd64 [installed]
ros-noetic-moveit-ros-warehouse/focal,focal,now 1.1.12-1focal.20230524.161231 amd64 [installed]
ros-noetic-moveit-ros/focal,focal,now 1.1.12-1focal.20230524.170111 amd64 [installed]
ros-noetic-moveit-servo/focal,focal,now 1.1.12-1focal.20230524.164354 amd64 [installed]
ros-noetic-moveit-setup-assistant/focal,focal,now 1.1.12-1focal.20230526.060324 amd64 [installed]
ros-noetic-moveit-simple-controller-manager/focal,focal,now 1.1.12-1focal.20230524.155333 amd64 [installed]
ros-noetic-moveit-visual-tools/focal,focal,now 3.6.1-1focal.20230524.161213 amd64 [installed]
ros-noetic-moveit/focal,focal,now 1.1.12-1focal.20230526.061512 amd64 [installed]
ros-noetic-pilz-industrial-motion-planner/focal,focal,now 1.1.12-1focal.20230524.165453 amd64 [installed]
ros-noetic-srdfdom/focal,focal,now 0.6.4-1focal.20230516.061359 amd64 [installed,automatic]
```
And these are the downgraded packages, without the issue:
```
ros-noetic-chomp-motion-planner/now 1.1.11-1focal.20230412.162323 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.155215]
ros-noetic-moveit-commander/now 1.1.11-1focal.20230412.171725 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.164353]
ros-noetic-moveit-core/now 1.1.11-1focal.20230412.160527 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.152908]
ros-noetic-moveit-fake-controller-manager/now 1.1.11-1focal.20230412.164501 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161352]
ros-noetic-moveit-kinematics/now 1.1.11-1focal.20230412.165026 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161147]
ros-noetic-moveit-msgs/focal,focal,now 0.11.4-1focal.20230216.003351 amd64 [installed]
ros-noetic-moveit-planners-chomp/now 1.1.11-1focal.20230412.171718 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.164352]
ros-noetic-moveit-planners-ompl/now 1.1.11-1focal.20230412.165157 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161205]
ros-noetic-moveit-planners/now 1.1.11-1focal.20230412.173800 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.170420]
ros-noetic-moveit-plugins/now 1.1.11-1focal.20230413.061611 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.162050]
ros-noetic-moveit-ros-benchmarks/now 1.1.11-1focal.20230412.164732 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.162901]
ros-noetic-moveit-ros-control-interface/now 1.1.11-1focal.20230413.061311 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.155610]
ros-noetic-moveit-ros-manipulation/now 1.1.11-1focal.20230412.170317 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.163019]
ros-noetic-moveit-ros-move-group/now 1.1.11-1focal.20230412.165613 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161815]
ros-noetic-moveit-ros-occupancy-map-monitor/now 1.1.11-1focal.20230412.162624 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.155332]
ros-noetic-moveit-ros-perception/now 1.1.11-1focal.20230412.164848 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161214]
ros-noetic-moveit-ros-planning-interface/now 1.1.11-1focal.20230412.170807 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.163543]
ros-noetic-moveit-ros-planning/now 1.1.11-1focal.20230412.162940 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.155710]
ros-noetic-moveit-ros-robot-interaction/now 1.1.11-1focal.20230412.164411 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161232]
ros-noetic-moveit-ros-visualization/now 1.1.11-1focal.20230412.171723 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.164446]
ros-noetic-moveit-ros-warehouse/now 1.1.11-1focal.20230412.164146 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.161231]
ros-noetic-moveit-ros/now 1.1.11-1focal.20230412.173509 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.170111]
ros-noetic-moveit-servo/now 1.1.11-1focal.20230412.171729 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.164354]
ros-noetic-moveit-setup-assistant/now 1.1.11-1focal.20230412.173508 amd64 [installed,upgradable to: 1.1.12-1focal.20230526.060324]
ros-noetic-moveit-simple-controller-manager/now 1.1.11-1focal.20230413.060342 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.155333]
ros-noetic-moveit-visual-tools/now 3.6.0-1focal.20230412.164641 amd64 [installed,upgradable to: 3.6.1-1focal.20230524.161213]
ros-noetic-moveit/now 1.1.11-1focal.20230413.061933 amd64 [installed,upgradable to: 1.1.12-1focal.20230526.061512]
ros-noetic-pilz-industrial-motion-planner/now 1.1.11-1focal.20230412.172843 amd64 [installed,upgradable to: 1.1.12-1focal.20230524.165453]
ros-noetic-srdfdom/now 0.6.3-1focal.20230215.215006 amd64 [installed,upgradable to: 0.6.4-1focal.20230516.061359]
```

--- a/moveit_core/robot_model/include/moveit/robot_model/robot_model.h
+++ b/moveit_core/robot_model/include/moveit/robot_model/robot_model.h
@@ -271,7 +271,14 @@ public:
    * will actually warp wrist (and all its descendants).
    */
   static const moveit::core::LinkModel* getRigidlyConnectedParentLinkModel(const LinkModel* link,
+                                                                           Eigen::Isometry3d& transform,
                                                                            const JointModelGroup* jmg = nullptr);
+  static const moveit::core::LinkModel* getRigidlyConnectedParentLinkModel(const LinkModel* link,
+                                                                           const JointModelGroup* jmg = nullptr)
+  {
+    Eigen::Isometry3d unused;
+    return getRigidlyConnectedParentLinkModel(link, unused, jmg);
+  }
 
   /** \brief Get the array of links  */
   const std::vector<const LinkModel*>& getLinkModels() const

--- a/moveit_core/robot_model/src/robot_model.cpp
+++ b/moveit_core/robot_model/src/robot_model.cpp
@@ -1265,10 +1265,13 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
   return nullptr;
 }
 
-const LinkModel* RobotModel::getRigidlyConnectedParentLinkModel(const LinkModel* link, const JointModelGroup* jmg)
+const LinkModel* RobotModel::getRigidlyConnectedParentLinkModel(const LinkModel* link, Eigen::Isometry3d& transform,
+                                                                const JointModelGroup* jmg)
 {
   if (!link)
     return link;
+
+  transform.setIdentity();
   const moveit::core::LinkModel* parent_link = link->getParentLinkModel();
   const moveit::core::JointModel* joint = link->getParentJointModel();
   decltype(jmg->getJointModels().cbegin()) begin{}, end{};
@@ -1289,6 +1292,7 @@ const LinkModel* RobotModel::getRigidlyConnectedParentLinkModel(const LinkModel*
 
   while (parent_link && is_fixed_or_not_in_jmg(joint))
   {
+    transform = link->getJointOriginTransform() * transform;
     link = parent_link;
     joint = link->getParentJointModel();
     parent_link = joint->getParentLinkModel();

--- a/moveit_core/robot_state/include/moveit/robot_state/attached_body.h
+++ b/moveit_core/robot_state/include/moveit/robot_state/attached_body.h
@@ -175,13 +175,6 @@ public:
    * The returned transform is guaranteed to be a valid isometry. */
   const Eigen::Isometry3d& getSubframeTransform(const std::string& frame_name, bool* found = nullptr) const;
 
-  /** \brief Get the fixed transform to a named subframe on this body (relative to the robot link)
-   *
-   * The frame_name needs to have the object's name prepended (e.g. "screwdriver/tip" returns true if the object's
-   * name is "screwdriver"). Returns an identity transform if frame_name is unknown (and set found to false).
-   * The returned transform is guaranteed to be a valid isometry. */
-  const Eigen::Isometry3d& getSubframeTransformInLinkFrame(const std::string& frame_name, bool* found = nullptr) const;
-
   /** \brief Get the fixed transform to a named subframe on this body, relative to the world frame.
    * The frame_name needs to have the object's name prepended (e.g. "screwdriver/tip" returns true if the object's
    * name is "screwdriver"). Returns an identity transform if frame_name is unknown (and set found to false).

--- a/moveit_core/robot_state/include/moveit/robot_state/robot_state.h
+++ b/moveit_core/robot_state/include/moveit/robot_state/robot_state.h
@@ -1379,7 +1379,7 @@ public:
    * This behaves the same as RobotModel::getRigidlyConnectedParentLinkModel,
    * but can additionally resolve parents for attached objects / subframes.
    *
-   * If transform is specified, return the relative transform from the returned parent link to frame.
+   * If transform is specified, return the (fixed) relative transform from the returned parent link to frame.
    */
   const moveit::core::LinkModel*
   getRigidlyConnectedParentLinkModel(const std::string& frame, Eigen::Isometry3d* transform = nullptr,

--- a/moveit_core/robot_state/src/cartesian_interpolator.cpp
+++ b/moveit_core/robot_state/src/cartesian_interpolator.cpp
@@ -160,10 +160,7 @@ double CartesianInterpolator::computeCartesianPath(RobotState* start_state, cons
     // Explicitly use a single IK attempt only: We want a smooth trajectory.
     // Random seeding (of additional attempts) would probably create IK jumps.
     if (start_state->setFromIK(group, pose * offset, link->getName(), consistency_limits, 0.0, validCallback, options))
-    {
-      start_state->update();
       traj.push_back(std::make_shared<moveit::core::RobotState>(*start_state));
-    }
     else
       break;
 

--- a/moveit_core/robot_state/src/robot_state.cpp
+++ b/moveit_core/robot_state/src/robot_state.cpp
@@ -1656,6 +1656,9 @@ bool RobotState::setFromIK(const JointModelGroup* jmg, const EigenSTL::vector_Is
   else if (consistency_limit_sets.size() == 1)
     consistency_limits = consistency_limit_sets[0];
 
+  // ensure RobotState is up-to-date before employing it in the IK solver
+  update(false);
+
   const std::vector<std::string>& solver_tip_frames = solver->getTipFrames();
 
   // Track which possible tips frames we have filled in so far

--- a/moveit_core/robot_state/src/robot_state.cpp
+++ b/moveit_core/robot_state/src/robot_state.cpp
@@ -809,36 +809,38 @@ const LinkModel* RobotState::getRigidlyConnectedParentLinkModel(const std::strin
     if (!hasAttachedBody(object))
       return nullptr;
     auto body{ getAttachedBody(object) };
-    if (!body->hasSubframeTransform(frame))
-      return nullptr;
+    bool found = false;
     if (transform)
-      *transform = body->getGlobalSubframeTransform(frame);
+      *transform = body->getSubframeTransform(frame, &found);
+    else
+      body->getSubframeTransform(frame, &found);
+    if (!found)
+      return nullptr;
+    if (transform)  // prepend the body transform
+      *transform = body->getPose() * *transform;
     link = body->getAttachedLink();
   }
   else if (hasAttachedBody(frame))
   {
     auto body{ getAttachedBody(frame) };
     if (transform)
-      *transform = body->getGlobalPose();
+      *transform = body->getPose();
     link = body->getAttachedLink();
   }
   else if (getRobotModel()->hasLinkModel(frame))
   {
     link = getLinkModel(frame);
     if (transform)
-    {
-      BOOST_VERIFY(checkLinkTransforms());
-      *transform = global_link_transforms_[link->getLinkIndex()];
-    }
+      transform->setIdentity();
+    if (!link)
+      return nullptr;
   }
   // link is valid and transform describes pose of frame w.r.t. global frame
-  auto* parent = getRobotModel()->getRigidlyConnectedParentLinkModel(link, jmg);
+  Eigen::Isometry3d link_transform;
+  auto* parent = getRobotModel()->getRigidlyConnectedParentLinkModel(link, link_transform, jmg);
   if (parent && transform)
-  {
-    BOOST_VERIFY(checkLinkTransforms());
-    // compute transform from parent link to frame
-    *transform = global_link_transforms_[parent->getLinkIndex()].inverse() * *transform;
-  }
+    // prepend link_transform to get transform from parent link to frame
+    *transform = link_transform * *transform;
   return parent;
 }
 

--- a/moveit_core/robot_state/test/robot_state_test.cpp
+++ b/moveit_core/robot_state/test/robot_state_test.cpp
@@ -686,7 +686,6 @@ TEST_F(OneRobot, rigidlyConnectedParent)
   EXPECT_EQ(robot_model_->getRigidlyConnectedParentLinkModel(link_b), link_a);
 
   moveit::core::RobotState state(robot_model_);
-  state.update();
 
   Eigen::Isometry3d a_to_b;
   EXPECT_EQ(state.getRigidlyConnectedParentLinkModel("link_b", &a_to_b), link_a);

--- a/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp
+++ b/moveit_ros/planning_interface/test/move_group_interface_cpp_test.cpp
@@ -345,6 +345,18 @@ TEST_F(MoveGroupTestFixture, JointSpaceGoalTest)
   testJointPositions(plan_joint_positions);
 }
 
+TEST_F(MoveGroupTestFixture, CartesianGoalTest)
+{
+  move_group_->setPoseReferenceFrame("world");
+  move_group_->setEndEffectorLink("panda_hand");
+  geometry_msgs::Pose pose;
+  pose.position.x = 0.417;
+  pose.position.y = 0.240;
+  pose.position.z = 0.532;
+  pose.orientation.w = 1.0;
+  EXPECT_TRUE(move_group_->setJointValueTarget(pose));
+}
+
 int main(int argc, char** argv)
 {
   ros::init(argc, argv, "move_group_interface_cpp_test");

